File: zad2.py
from heapq import heappop, heappush

def queue_peoples(visitors, num_windows):
    window_times = [0] * num_windows
    windows = [[] for _ in range(num_windows)]

    visitors = [(-time, -talon) for time, talon in visitors]
    heapq.heapify(visitors)

    while visitors:
        time, talon = heappop(visitors)
        min_window_index = window_times.index(min(window_times))
        windows[min_window_index].append(-talon)
        window_times[min_window_index] -= time

    for i, window in enumerate(windows):
        print(f"Окно {i+1} ({window_times[i]} минут):", ' '.join(f"T{-talon}" for talon in window))


peoples = []
print(">>> Введите кол-во рабочих окон:")
num_windows = int(input("<<< "))

print(">>> Для получения талона введите: ENQUEUE и время посещения\n\tДля формирования очереди введите: DISTRIBUTE\n\t\tДля выхода введите: EXIT")
talon = 100
while True:
    command = input("<<< ")
    if command == "ENQUEUE":
        time = int(input())
        talon += 10
        print(f">>> T{talon}")
        heappush(peoples, (-time, -talon))
    elif command == "DISTRIBUTE":
        queue_peoples(peoples, num_windows)
    elif command == "EXIT":
        print(">>> Выход...")
        break
    else:
        print(">>> Команда введена неверно")

File: zad2.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

void queuePeoples(vector<pair<int, int>>& visitors, int numWindows) {
    vector<int> windowTimes(numWindows, 0);
    vector<int> windowNumbers(numWindows);
    vector<vector<int>> WINDOWS(numWindows);
    // Работа с номерами окон
    for (int i = 0; i < numWindows; ++i) {
        windowNumbers[i] = i + 1;
    }

    int numNextPeople = 0; // Номер следующего посетителя

    while (!visitors.empty()) {
        // Находим посетителя с самым большим временем посещения
        auto maxPeopleIndex = max_element(visitors.begin(), visitors.end()) - visitors.begin();

        // Находим индекс окна с минимальным временем работы
        auto minWindowIndex = min_element(windowTimes.begin(), windowTimes.end()) - windowTimes.begin();

        WINDOWS[minWindowIndex].push_back(visitors[maxPeopleIndex].second);

        
        windowTimes[minWindowIndex] += visitors[maxPeopleIndex].first; // Увеличиваем время работы окна на время обслуживания посетителя

        visitors.erase(visitors.begin() + maxPeopleIndex); // Удаляем посетителя из очереди

        numNextPeople++;
    }

    for (int i = 0; i < WINDOWS.size(); ++i) {
        cout << "Окно " << i+1 << "(" << windowTimes[i] << " минут): ";
        for (auto elem : WINDOWS[i]) {
            cout << "T" << elem << " ";
        }
        cout << endl;
    }

}

int main(){
    setlocale(LC_ALL, "Rus");

    vector<pair<int, int>> Peoples; // вектор пара, в котором будут хранится время и талоны посетителей
    cout << ">>> Введите кол-во рабочих окон:" << endl;
    cout << "<<< ";
    int NumsWindows;
    cin >> NumsWindows;

    cout << ">>> Для получения талона введите: ENQUEUE и время посещения\n\tДля формирования очереди введите: DISTRIBUTE\n\t\tДля выхода введите: EXIT" << endl;
    string command;
    bool k = true; // условие выхода из while
    int time; // время посетителя
    int talon = 100; // талон посетителя
    while (k == true) {
        cout << "<<< ";
        cin >> command;
        if (command == "ENQUEUE") {
            cin >> time;
            talon += 10;
            cout << ">>> T" << talon << endl;
            Peoples.push_back({time, talon});
        } else if (command == "DISTRIBUTE") {
            queuePeoples(Peoples, NumsWindows); // ф-ия формирования очереди
        } else if (command == "EXIT") {
            cout << ">>> Выход..." << endl;
            k = false;
        } else {
            cout << ">>> Команда введена неверно" << endl;
        }
    }

    return 0;
}
